src/passes/OptimizeInstructions.cpp

```diff
@@ -1345,15 +1381,58 @@ struct OptimizeInstructions
         }
       }
       // operations on all 1s
-      // TODO: shortcut method to create an all-ones?
       if (right->value == Literal(int32_t(-1)) ||
           right->value == Literal(int64_t(-1))) {
         if (binary->op == Abstract::getBinary(type, Abstract::And)) {
+          // x & -1   ==>   x
           return binary->left;
         } else if (binary->op == Abstract::getBinary(type, Abstract::Or) &&
                    !EffectAnalyzer(getPassOptions(), features, binary->left)
                       .hasSideEffects()) {
+          // x | -1   ==>   -1
           return binary->right;
+        } else if (binary->op == Abstract::getBinary(type, Abstract::Sub)) {
+          // x - (-1)   ==>   x + 1
+          binary->op = Abstract::getBinary(type, Abstract::Add);
+          right->value = Literal::makeFromInt32(1, type);
+          return binary;
+        } else if (binary->op == Abstract::getBinary(type, Abstract::RemS) &&
+                   !EffectAnalyzer(getPassOptions(), features, binary->left)
+                      .hasSideEffects()) {
+          // (signed)x % -1     ==>   0
+          right->value = Literal::makeSingleZero(type);
+          return right;
+        } else if (binary->op == Abstract::getBinary(type, Abstract::GtU) &&
+                   !EffectAnalyzer(getPassOptions(), features, binary->left)
+                      .hasSideEffects()) {
+          // (unsigned)x > -1   ==>   0
+          right->value = Literal::makeSingleZero(Type::i32);
+          right->type = Type::i32;
+          return right;
+        } else if (binary->op == Abstract::getBinary(type, Abstract::LtU)) {
+          // (unsigned)x < -1   ==>   x != -1
+          // friendlier to JS emitting as we don't need to write an unsigned
+          // -1 value which is large.
+          binary->op = Abstract::getBinary(type, Abstract::Ne);
+          return binary;
+        } else if (binary->op == Abstract::getBinary(type, Abstract::DivU)) {
+          // (unsigned)x / -1   ==>   x == -1
+          binary->op = Abstract::getBinary(type, Abstract::Eq);
+          binary->type = Type::i32;
+          return binary;
```

Comment:
@kripken fuzzer found bug here and I'm stumped. I tried different approaches like create binary from scratch via Builder but always the same problem:
```
expected i64, seeing i32 from
0x7f7f9d902e10
Assertion failed: (Type::isSubType(type, curr->type)), function visit, file /Volumes/Archive/Projects/Github/binaryen/src/wasm-interpreter.h, line 204.
!
-----------------------------------------
Exception:
  File "./scripts/fuzz_opt.py", line 791, in <module>
  File "./scripts/fuzz_opt.py", line 610, in test_one
  File "./scripts/fuzz_opt.py", line 73, in run
  File "/Applications/Xcode.app/Contents/Developer/Library/Frameworks/Python3.framework/Versions/3.7/lib/python3.7/subprocess.py", line 395, in check_output
    **kwargs).stdout
  File "/Applications/Xcode.app/Contents/Developer/Library/Frameworks/Python3.framework/Versions/3.7/lib/python3.7/subprocess.py", line 487, in run
    output=stdout, stderr=stderr)
-----------------------------------------
!
-6
```

Seed num: `13377051698656758529`

Reply:
I figured out what the problem was. We can't do `x / -1   ==>   x == -1` if `x` is 64-bit value. In this case we also should upcast result to `i64` via `i64.extend_i32_s`. This will increase size so now it applies only for `DivUInt32`